File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/17-Graphs-BFS/LeetCode/LeetCode-0515.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> largestValues(TreeNode* root) {
        vector<int> result;
        if (root == NULL) return result;
        
        queue<TreeNode *> Q;
        Q.push(root);
                    
        while (!Q.empty()){
            int n = Q.size();  // # of nodes at this level
            
            // Add a new list to the end of the result
            int maxVal = INT_MIN;
            for (int i=0; i<n; i++){
                auto p = Q.front();
                Q.pop();
                maxVal = std::max(maxVal, p->val);
                if (p->left) Q.push(p->left);
                if (p->right) Q.push(p->right);                    
            } // end-for
            result.push_back(maxVal);
        } //end-while        

        return result;        
    } // end-largestValues
};

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/17-Graphs-BFS/LeetCode/LeetCode-0515.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def largestValues(self, root: Optional[TreeNode]) -> List[int]:
        result = []
        if root is None: return result

        Q = []
        Q.append(root)
        
        while len(Q) > 0:
            n = len(Q)  # Number of nodes at this level
            
            # Add a new list to the end of the result
            maxVal = -1<<31
            for i in range(n):
                p = Q.pop(0)
                maxVal = max(maxVal, p.val)
                if p.left is not None: Q.append(p.left)
                if p.right is not None: Q.append(p.right)
            result.append(maxVal)
        return result        